src/main/java/com/xjeffrose/xio/http/Client.java

```diff
@@ -0,0 +1,104 @@
+import io.netty.handler.ssl.SslContext;
+import java.net.InetSocketAddress;
+import java.util.ArrayDeque;
+import java.util.Queue;
+import io.netty.channel.ChannelHandler;
+import com.xjeffrose.xio.core.XioMessageLogger;
+import com.xjeffrose.xio.core.XioIdleDisconnectHandler;
+import java.util.function.Supplier;
+import java.util.concurrent.Future;
+import io.netty.channel.ChannelFutureListener;
+
+public class Client {
+
+  private final ClientState state;
+  private final Supplier<ChannelHandler> appHandler;
+  Channel channel;
+
+  public Client(ClientState state, Supplier<ChannelHandler> appHandler) {
+    this.state = state;
+    this.appHandler = appHandler;
+  }
+
+  private ChannelHandler buildHttp2Handler() {
+    return new Http2HandlerBuilder().server(false).build();
+  }
+
+  private ChannelFuture connect() {
+    Bootstrap b = new Bootstrap();
+    b.channel(state.channelConfig.channel());
+    b.group(state.channelConfig.workerGroup());
+    b.handler(
+        new ChannelInitializer() {
+          public void initChannel(Channel channel) {
+            if (state.sslContext != null) {
+              channel
+                  .pipeline()
+                  .addLast(
+                      "ssl handler",
+                      state.sslContext.newHandler(
+                          channel.alloc(), state.remote.getHostString(), state.remote.getPort()));
+            }
+            channel
+                .pipeline()
+                .addLast(
+                    "negotiation handler",
+                    new HttpClientNegotiationHandler(Client.this::buildHttp2Handler))
+                .addLast("codec", CodecPlaceholderHandler.INSTANCE)
+                .addLast("application codec", ApplicationCodecPlaceholderHandler.INSTANCE)
+                .addLast("idle handler", new XioIdleDisconnectHandler(60, 60, 60))
+                .addLast("message logging", new XioMessageLogger(Client.class, "objects"))
+                .addLast("request buffer", new RequestBuffer())
+                .addLast("app handler", appHandler.get());
+            //              .addLast(new RawBackendHandler(ctx));
+          }
+        });
+
+    return b.connect(state.remote);
+  }
+
+  private void connected(ChannelFuture f) {
+    if (f.isDone() && f.isSuccess()) {
+      // log?
+    } else {
+      // log?
+      throw new RuntimeException(f.cause());
+    }
+  }
+
+  public ChannelFuture write(Request request) {
+    if (channel == null) {
+      ChannelFuture future = connect();
+      ChannelFutureListener l =
+          f -> {
+            if (f.isDone() && f.isSuccess()) {
+              // log?
+            } else {
+              // log?
+              // fail the write future?
+              throw new RuntimeException(f.cause());
+            }
+          };
+      channel = future.channel();
+      // connect().addListener(f -> connected((ChannelFuture) f));
+      connect().addListener(l);
+    }
+    return channel.writeAndFlush(request);
```

Comment:
This is a goofy netty thing.

> Returns a channel where the I/O operation associated with this future takes place.

In my experience this means you can get the `Channel` from the `ChannelFuture` as soon as it's `connect()`ing or `bind()`ing or what have you. That being said, this works and is thread safe as long as the thread on the `EventLoopGroup` that this channel is associated with is the same thread that's invoking `write`. I'm guaranteeing that right now. In the future we need a more complex setup that chains the futures like so:
```
connect().andThen(write).andThen(write)
```

Netty `ChannelFutureListener` doesn't support this syntax, but you can do something close to it. We'll also want to bundle the initial connect and write promises together since the method only returns a single `ChannelFuture`